pennylane/interfaces/execution.py

```diff
@@ -360,6 +360,22 @@ def cost_fn(params, x):
         for i, tape in enumerate(tapes):
             tapes[i] = expand_fn(tape)
 
+        if gradient_kwargs.get("method", "") == "adjoint_jacobian":

```

Comment:
This performs further expansion for adjoint differentiation. It makes sure all trainable operations have a generator.

This code should be moved elsewhere long-term, but it fixes the problem for now.

Reply:
@albi3ro I'm not sure what this is addressing, since the PR title/description don't describe anything re: adjoint. Is this potentially something out of scope, that should be a separate PR?

> This code should be moved elsewhere long-term, but it fixes the problem for now.

This should go in the `transforms/tape_expand.py` file. Any reason for not doing it in this PR?

Reply:
The changes were breaking adjoint jacobian.

I can't add the expansion to `QubitDevice.adjoint_jacobian`, because that method does not have access to differentiability information.  We have to perform this expansion prior to removing differentiability information.

I had to stick this information somewhere we had access to differentiability information.

While I could extra the below logic into a helper method that is called here, I am trying to implement a quick patch that will be easy to remove or modify later.

Reply:
> I can't add the expansion to `QubitDevice.adjoint_jacobian`, because that method does not have access to differentiability information. We have to perform this expansion prior to removing differentiability information.

Definitely something we need to address in a new Device API.

> While I could extra the below logic into a helper method that is called here, I am trying to implement a quick patch that will be easy to remove or modify later.

I defer to the technical team (@antalszava likely has thoughts here regarding best practices), but it sounds like a relatively small change (moving the logic to where similar logic is already stored) with significant advantages:

- reduction in long term technical debt
- ability to better document it via docstring
- unit testing, since all that appears at the moment are integration tests

Reply:
We have a choice here.  We can ship this feature with the release with a short-term increase to technical debt, or we can leave this out and not provide device support for operator arithmetic.

I really, really want to improve this entire infrastructure from the ground up as it is rife with technical debt and badly written, confusing code, but shipping is a feature.  I was told to implement the quick patch (though this still took quite a bit of work to figure out) rather than wait for an ADR to get approved and implemented.

Reply:
@albi3ro sorry I think there is some miscommunication on my end, I don't see a need to go through any design or ADR process here, or fix existing technical debt :)

I think the hotfix for this logical edge case makes sense, I'm just suggesting in this PR it would be a better fit in a different file/location, that currently contains this logic.

Reply:
I think if we leave all of this logic in the same place, it will be easier to remove later. If we make a private method and then call the private method here, we will have to delete code in multiple places.

Reply:
No strong suggestions from my side on this, other than adding the word `TODO` into the comment. I think so long as we have a comment in the codebase about the planned removal, we should be in the clear.

This assumes that:
* We will very soon remove this code (i.e., next release, 2 releases max.);
* The same code is not used anywhere else in the code base;
* All the changes are developer facing and it is potentially beneficial for developers to see the vicinity of the code base to see how this block fits.

Reply:
Would the hotfix only be removed after when we can move the logic into `QubitDevice.adjoint_jacobian` and so with the new Device API?

Reply:
I want to extract the tape expansion and support checking from the device into a new object owned by a device.  This was have a minimal, generic, standard interface. Tools will be provided to implement this interface easily on different devices. For example, it could be constructed from a list of string names.

One thing we need this infrastructure to do is allow for gradient-specific device support logic.  This includes adjoint differentiation (which has been requested by the performance team) and backpropagation, if not more.

Reply:
Great, thank you for the explanation.

It seems like there are multiple steps before an alternative can be rolled out:
1. Work on the design;
2. Discuss and have it accepted by the team;
3. Apply the implementation.

This could take some time and attention from multiple stakeholders, so perhaps it's best to consider the hotfix as something that's here to stay.

* Are there any blockers to moving this code to the `transforms/tape_expand.py` file and importing a function here?
* Also, could the `create_expand_fn` function be used in that file?

It seems like that using `create_expand_fn`, we could transform the tapes here (might be missing something though).

Reply:
Feel free to branch off this and propose an alternate solution. I can't think of a simple, quick solution myself, but that doesn't mean one doesn't exist.

I don't want to distribute this patch over multiple parts of the code base. That will make it much harder to remove.

In looking at `create_expand_fn`,using it would *add* complexity. How much code would we actually be abstracting away at the cost of adding in a new function you have to remember?

Things that would actually make this code simpler:

* Allow `qml.operation.is_trainable` to work on jax parameters
* Have separate stopping conditions for measurements and operations

Currently, we have to have `~qml.operation.is_measurement` to prevent errors from being raised when the stopping condition is applied to a measurement.

Reply:
https://github.com/PennyLaneAI/pennylane/pull/2879

> In looking at create_expand_fn,using it would add complexity.

Why?

> How much code would we actually be abstracting away at the cost of adding in a new function you have to remember?

No need to remember, it is used in the execute file and placed in a place where every other expansion function is.